### PR TITLE
Initial version of the Tree of Signature

### PR DESCRIPTION
**Pull request by beurdouche**

### CODE REVIEW

**Comment by bifurcation** [comment]
It seems like there's a major thing missing here, namely how the sender of a Commit transmits the new node signatures to the other members of the group.  We need something like that before this lands.

The most obvious approach would be to add a `signature` field to DirectPathNode, as you have with RatchetNode.

**Comment by bifurcation** [comment]
It would also be handy to include in the intermediate node an indication of which leaf signed it.  Otherwise you have to do "trial verifications" (cf. trial decryption), which is expensive.

File: draft-ietf-mls-protocol.md [hunk 1]
```diff
@@ -1960,16 +1983,11 @@ keys for nodes in its direct path that are being reset by the corresponding
...
-    HPKEPublicKey public_key;
-    opaque unmerged_leaves<0..2^32-1>;
-    optional<ClientInitKey> client_init_key;
-} RatchetNode;
-
 struct {
   // GroupContext inputs
   opaque group_id<0..255>;
   uint64 epoch;
+  opaque root_node_hash<0..255>;
```
Comment (raphaelrobert) [inline draft-ietf-mls-protocol.md]: This is a good starting point, we should see if we need something fancier in the future, like a Merkle tree (or reuse the node hashes).

File: draft-ietf-mls-protocol.md [hunk 2]
```diff
@@ -1089,40 +1100,52 @@ struct {
...
+~~~~~
 struct {
     uint32 node_index;
-    optional<ParentNodeInfo> info;
+    optional<ParentNode> parent_node;
     opaque left_hash<0..255>;
     opaque right_hash<0..255>;
-} ParentNodeHashInput;
+    opaque signature<0..2^16-1>;
+} ParentNodeHash;
```
Comment (raphaelrobert) [inline draft-ietf-mls-protocol.md]: Isn't this where we discussed that the signer's index is missing?

**Comment by kelrit** [comment]
> It seems like there's a major thing missing here, namely how the sender of a Commit transmits the new node signatures to the other members of the group. We need something like that before this lands.
> 
> The most obvious approach would be to add a `signature` field to DirectPathNode, as you have with RatchetNode.

Was this addressed? it's mentioned that the signature should be transmitted in "Synchronizing Views of the Tree" but there's no signature in the transmitted DirectPathNode

**Comment by Bren2010** [comment]
@kelrit #287 changes the strategy for signing the tree, everything should be sorted out there.

**Pull request merged**